index.js

```diff
@@ -115,9 +115,7 @@ function toRunCommand (inspectObj, name) {
   if (cfg.ExposedPorts) {
     rc = appendObjectKeys(rc, '--expose', cfg.ExposedPorts)
   }
-  rc = appendArray(rc, '-e', cfg.Env, (env) => {
-    return /\s/g.test(env) ? '"' + env + '"' : env
-  })
+  rc = appendArray(rc, '-e', cfg.Env, (env) => '\'' + env.replace(/'/g, '\\\'') + '\'')
```

Comment:
Based on my testing and on the example given in the aforementioned [SO answer](https://stackoverflow.com/a/20053121/1174467):

```
'I'\''m a s@fe $tring which ends in newline
'
```

I think we need line 118 to be this instead:

```js
rc = appendArray(rc, '-e', cfg.Env, (env) => '\'' + env.replace(/'/g, '\'\\\'\'') + '\'')
```

because we need to replace each internal `'` character with `'\''` in order to be syntax-compatible with bash and zsh.

You can test this with the following commands. The first is based on output generated from this PR branch, and the command fails in bash and zsh. The second is based on output with my recommended changes, and it succeeds in both shells.

```console
$ docker run --name competent_agnesi --restart no -h 3936b9c3a94b -e 'special_char_env_var2=abc(\')123' -e 'PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin' -a stdout -a stderr tianon/true /true
-bash: syntax error near unexpected token `)'
```

```console
$ docker run --name competent_agnesi --restart no -h 3936b9c3a94b -e 'special_char_env_var2=abc('\'')123' -e 'PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin' -a stdout -a stderr tianon/true /true
```

Inspecting the generated container confirms that the env var is set properly to `special_char_env_var2=abc(')123`.

Reply:
sorry, agreed this was wrong, fixed in cf00449